src/decoder/cuda-decoder-utils.cu

```diff
@@ -0,0 +1,237 @@
+
+  cudaMemcpyAsync(final_d, final_h, sizeof(float)*numStates, cudaMemcpyHostToDevice,
+                  cudaStreamPerThread);
+  cudaMemcpyAsync(e_offsets_d, e_offsets_h, sizeof(unsigned int) * (numStates + 1),
+                  cudaMemcpyHostToDevice, cudaStreamPerThread);
+  cudaMemcpyAsync(ne_offsets_d, ne_offsets_h,
+                  sizeof(unsigned int) * (numStates + 1), cudaMemcpyHostToDevice,
+                  cudaStreamPerThread);
+
+  // although using managed memory in the following, we still have a CPU
+  // copy to speed up the CPU lattice processing
+  cudaMallocHost(&arc_weights_h, arc_count * sizeof(BaseFloat));
+  cudaMallocHost(&arc_nextstates_h, arc_count * sizeof(StateId));
+  cudaMallocHost(&arc_ilabels_h, arc_count * sizeof(int32));
+  cudaMallocHost(&arc_olabels_h, arc_count * sizeof(int32));
+
+  // to reduce memory usage, we use cudaMallocManaged, which doesn't
+  // allocate in GPU at once
+  cuda_malloc_managed_preferred_device((void**)&arc_weights_d,
+                                       arc_count * sizeof(BaseFloat));
+  bytes_cudaMalloc += arc_count * sizeof(BaseFloat);
+  cuda_malloc_managed_preferred_device((void**)&arc_nextstates_d,
+                                       arc_count * sizeof(StateId));
+  bytes_cudaMalloc += arc_count * sizeof(StateId);
+  cuda_malloc_managed_preferred_device((void**)&arc_ilabels_d,
+                                       arc_count * sizeof(int32));
+  bytes_cudaMalloc += arc_count * sizeof(int32);
+  cuda_malloc_managed_preferred_device((void**)&arc_olabels_d,
+                                       arc_count * sizeof(int32));
+  bytes_cudaMalloc += arc_count * sizeof(int32);
+
+  // now populate arc data
+  int e_idx = 1;        // save room for dummy arc (so start at 1)
+  int ne_idx = e_count + 1; // starts where e_offsets ends
+
+  // create dummy arc
+  arc_weights_h[0] = StdWeight::One().Value();
+  arc_nextstates_h[0] = fst.Start();
+  arc_ilabels_h[0] = 0;
+  arc_olabels_h[0] = 0;
+
+  for (int i = 0; i < numStates; i++) {
+    // count emmiting and non_emitting arcs
+    for (fst::ArcIterator<fst::Fst<StdArc> > aiter(fst, i); !aiter.Done();
+         aiter.Next()) {
+      StdArc arc = aiter.Value();
+      int idx;
+      if (arc.ilabel != 0) { // emitting
+        idx = e_idx++;
+      } else {
+        idx = ne_idx++;
+      }
+      arc_weights_h[idx] = arc.weight.Value();
+      arc_nextstates_h[idx] = arc.nextstate;
+      arc_ilabels_h[idx] = arc.ilabel;
+      arc_olabels_h[idx] = arc.olabel;
+    }
+  }
+
+  memcpy(arc_weights_d, arc_weights_h, arc_count * sizeof(BaseFloat));
+  memcpy(arc_nextstates_d, arc_nextstates_h, arc_count * sizeof(StateId));
+  memcpy(arc_ilabels_d, arc_ilabels_h, arc_count * sizeof(int32));
+  memcpy(arc_olabels_d, arc_olabels_h, arc_count * sizeof(int32));
+
+  cudaStreamSynchronize(cudaStreamPerThread);
+  POP_RANGE
+}
+
+void CudaFst::Finalize() {
+  PUSH_RANGE("CudaFst destructor", 0);
+  cudaFreeHost(final_h);
```

Comment:
Please wrap the cuda API calls with the CUDA_CALL macro.

Reply:
done